app/models/observation.rb

```diff
@@ -193,6 +239,164 @@ class Observation < AbstractModel
+      joins(:namings).where(namings: { name_id: name_ids })
+    elsif args[:of_look_alikes]
+      joins(:namings).where(namings: { name_id: name_ids }).
+        where.not(name: name_ids)
+    else
+      where(name_id: name_ids)
+    end
+  }
+  scope :of_name_like,
+        ->(name) { where(name: Name.text_name_includes(name)) }
+  scope :by_user, ->(user) { where(user: user) }
+  scope :with_location, -> { where.not(location: nil) }
+  scope :without_location, -> { where(location: nil) }
+  scope :at_location, ->(location) { where(location: location) }
+  scope :in_region,
+        ->(where) { where(Observation[:where].matches("%#{where}")) }
+  scope :in_box, # Use named parameters (n, s, e, w), any order
+        lambda { |**args|
+          if args[:s]&.between?(0, 90) && args[:n]&.between?(0, 90) &&
+             args[:w]&.between?(-180, 180) && args[:e]&.between?(-180, 180) &&
+             args[:s] <= args[:n] &&
+             ((args[:w] <= args[:e]) || (args[:w] >= 0 && args[:e] <= 0))
+
+            # expand box by epsilon to create leeway for Float rounding
+            # Fixes a bug where Califoria fixture was not in a box
+            # defined by the fixture's north, south, east, west
+            epsilon = 0.00001
+            n = args[:n] + epsilon
+            s = args[:s] - epsilon
+            e = args[:e] + epsilon
+            w = args[:w] - epsilon
+
+            # Does box straddle 180 deg?
+            if args[:e] < args[:w]
+              where(
+                # Observation[:long] between w & 180 OR between 180 and e
+                (Observation[:lat] >= s).
+                and(Observation[:lat] <= n).
+                and(Observation[:long] >= w).
+                or(Observation[:long] <= e)
+              )
+            else
+              where(
+                (Observation[:lat] >= s).
+                and(Observation[:lat] <= n).
+                and(Observation[:long] >= w).
+                and(Observation[:long] <= e)
+              )
+            end
+          else
+            Observation.none
+          end
+        }
+  scope :is_collection_location, -> { where(is_collection_location: true) }
+  scope :not_collection_location, -> { where(is_collection_location: false) }
+  scope :with_image, -> { where.not(thumb_image: nil) }
+  scope :without_image, -> { where(thumb_image: nil) }
+  scope :with_notes, -> { where.not(notes: Observation.no_notes) }
+  scope :without_notes, -> { where(notes: Observation.no_notes) }
+  scope :has_notes_field,
+        ->(field) { where(Observation[:notes].matches("%:#{field}:%")) }
+  scope :notes_include,
+        ->(notes) { where(Observation[:notes].matches("%#{notes}%")) }
+  scope :with_specimen, -> { where(specimen: true) }
+  scope :without_specimen, -> { where(specimen: false) }
+  scope :with_sequence, -> { joins(:sequences).distinct }
+  scope :without_sequence, -> { missing(:sequences) }
+  # TODO: Please check if this is how we do confidence.
```

Comment:
I don't know how we do confidence; I've always avoided that part of the code.
(I guess it's time to try to learn.)
Suggestion: 
Instead have `scope: confidence` like the pattern_search `confidence`.
(Takes a single number or range from -100 through +100. )
See  https://docs.google.com/document/d/1pUdpDL2_GzBxB8O-vO8WBYDTl2jQJrQ7Q0Q0k4GGU8o/edit#heading=h.2aopncqntbog
If you need `< x` then use `where.not.confidence(x..100)` similar for `> x`.
Or add scopes that chain `confidence`.

Reply:
Thanks Jason! I'm looking through the code now, and that's a big help.

> Confidence has always been a synonym for observations.vote_cache which is between -3 and 3, and I think for the UI I scale it up to -100 to 100.
> […](#)
> On Fri, Jul 29, 2022 at 7:18 PM Joseph D. Cohen ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In app/models/observation.rb <[#1082 (comment)](https://github.com/MushroomObserver/mushroom-observer/pull/1082#discussion_r933680155)> : > + } + scope :is_collection_location, -> { where(is_collection_location: true) } + scope :not_collection_location, -> { where(is_collection_location: false) } + scope :with_image, -> { where.not(thumb_image: nil) } + scope :without_image, -> { where(thumb_image: nil) } + scope :with_notes, -> { where.not(notes: Observation.no_notes) } + scope :without_notes, -> { where(notes: Observation.no_notes) } + scope :has_notes_field, + ->(field) { where(Observation[:notes].matches("%:#{field}:%")) } + scope :notes_include, + ->(notes) { where(Observation[:notes].matches("%#{notes}%")) } + scope :with_specimen, -> { where(specimen: true) } + scope :without_specimen, -> { where(specimen: false) } + scope :with_sequence, -> { joins(:sequences).distinct } + scope :without_sequence, -> { missing(:sequences) } + # TODO: Please check if this is how we do confidence. I don't know how we do confidence; I've always avoided that part of the code. (I guess it's time to try to learn.) Suggestion: Instead have scope: confidence like the pattern_search confidence. (Takes a single number or range from -100 through +100. ) See https://docs.google.com/document/d/1pUdpDL2_GzBxB8O-vO8WBYDTl2jQJrQ7Q0Q0k4GGU8o/edit#heading=h.2aopncqntbog If you need < x then use where.not.confidence(x..100) similar for > x. Or add scopes that chain confidence. — Reply to this email directly, view it on GitHub <[#1082 (review)](https://github.com/MushroomObserver/mushroom-observer/pull/1082#pullrequestreview-1056143281)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAYTNNPMT7TW34C43XJCPT3VWRRFZANCNFSM53JGSGUA> . You are receiving this because you were mentioned.Message ID: ***@***.*** com>